### PR TITLE
fix(raft): member type can be updated reliably

### LINKED ISSUE #14548: `RaftTest.testDemoteLeader` is unstable
**Summary**

> Try to answer the following as best as possible

- How often does the test fail? Once that I know of
- Does it block your work? No
- Do we suspect that it is a real failure? Probably not but we should investigate

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Error:    Run 1: RaftTest.testDemoteLeader:238->ConcurrentTestCase.await:110 ? Timeout Test timed out while waiting for an expected result, expectedResumes: 2, actualResumes: 1
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**

[Flaky RaftTest.zip](https://github.com/camunda/zeebe/files/12765358/Flaky.RaftTest.zip)

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This fixes https://github.com/camunda/zeebe/issues/14548

While these bugs could have been very bad, they were only triggered when changing a members type which we only do in tests.

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java [hunk 1]
```diff
@@ -71,10 +77,28 @@ public boolean requiresJointConsensus() {
...
-    all.addAll(oldMembers);
-    return all;
+    return new HashSet<>(
+        Stream.concat(newMembers.stream(), oldMembers.stream())
+            .collect(
+                Collectors.toMap(
+                    RaftMember::memberId,
+                    Function.identity(),
+                    BinaryOperator.maxBy(Comparator.comparing(RaftMember::getType))))
+            .values());
```
Comment (oleschoenburg) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:93]: :thought_balloon: Happy to get suggestions on how to do this more elegantly...

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java [hunk 2]
```diff
@@ -71,10 +77,28 @@ public boolean requiresJointConsensus() {
...
-    all.addAll(newMembers);
-    all.addAll(oldMembers);
-    return all;
+    return new HashSet<>(
+        Stream.concat(newMembers.stream(), oldMembers.stream())
+            .collect(
+                Collectors.toMap(
+                    RaftMember::memberId,
+                    Function.identity(),
+                    BinaryOperator.maxBy(Comparator.comparing(RaftMember::getType))))
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:92]: :question: Is this comparing by ordinal?

File: atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java [hunk 3]
```diff
@@ -376,6 +373,79 @@ void quorumWhenClusterWasSingleMember() {
...
+            1,
+            Instant.now().toEpochMilli(),
+            List.of(
+                localMember,
+                new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+                new DefaultRaftMember(new MemberId("3"), Type.PASSIVE, Instant.now())),
+            List.of());
+    context.configure(newConfiguration);
+
+    // then -- member 3 is no longer a voting member
```
Comment (npepinpe) [inline atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java:441]: :wrench: comment is sort of misleading, I assume was copy pasted :smile:

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java (hunk 1 above)
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:93]: Maybe I'm old, but I would just loop over each collection in turn. I wouldn't say it's more elegant though, this is fine.

**Review by npepinpe — APPROVED** [review]
:rocket:

File: atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java (hunk 2 above)
Comment (oleschoenburg) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/system/Configuration.java:92]: Yes, the type is an enum and inherits the default `Comparable` from enums. We already rely on the ordering of enum variants in several other places so I think that's okay.

File: atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java (hunk 3 above)
Comment (oleschoenburg) [reply atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java:441]: Oops yeah :+1:

**Comment by oleschoenburg** [comment]
bors merge

**Pull request merged**